### ISSUE TITLE
AbstractMapVisualizer state is spammed when loading cached tiles

### ISSUE THREAD

**Issue by DennyLindberg** [mapbox/mapbox-unity-sdk#896]
_Note: If this is a bug or support ticket, please provide the following information:_
* **Unity version:** 2017.4.4f1
  * **Scripting Runtime Version:** Stable (.NET 3.5 Equivalent)
  * **Scripting Backend:** Mono
  * **Api Compatibility Level:** .NET 2.0 Subset
* **Mapbox SDK version:** 1.4.3
* **The platform you're building to:** Android
* **A description of what you're trying to do:**

In AbstractMapVisualizer.cs there is an action called OnMapVisualizerStateChanged. This callback behaves differently depending on if tiles are cached or not. Let me do a rundown of the current code execution flow and try to point out the problem.

OnMapVisualizerStateChanged is tied to the State variable, which is updated in relation to the factories as they process' new tiles from the TileProvider. If I step through the chain of events in the code it looks something like the following:

TileProvider.AddTile() -> OnTileAdded() -> AbstractMap.TileProvider_OnTileAdded() -> AbstractMapVisualizer.LoadTile() -> AbstractTileFactory.Register() -> Subclasses of AbstractTileFactory.OnRegistered() -> DataFetcher -> Creates a RasterTile -> Which calls MapboxAccess.Request() to get raster data -> CachingWebFileSource.Request() then either asynchronously gets the data from the server **or returns what it has directly**

Which can be seen with this code snippet
```
// immediately return cached tile
callback(Response.FromCache(cachedItem.Data));
```

Now to the problem. When the TileProvider updates or initializes it will add several tiles through a typical synchronous loop. As each tile is added it will trickle down the long chain above. On a first run of the program the raster data must be fetched from the server, which will cause a delay between when the tile is added and when the factory has finished loading. This will cause an offset between the number of tiles added and number of factories running and thus the state of the visualizer will be stable.

However, if there is an existing cache the state of the visualizer will be changed directly after a tile is added and thus each factory instantly returns.

* **Steps to recreate the bug if appropriate:**

Open ZoomableMap. Add a script component containing the code below. (it will log a message whenever a tile is added or the visualizer state changes)

```
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

using Mapbox.Unity.Map;
using Mapbox.Map;

public class BugReport : MonoBehaviour 
{
	void Awake() 
	{
		AbstractMap map = FindObjectOfType<AbstractMap>();

		if (map != null)
		{
			map.OnInitialized += () => {
				AbstractMapVisualizer visualizer = FindObjectOfType<AbstractMapVisualizer>();
				AbstractTileProvider tileProvider = FindObjectOfType<AbstractTileProvider>();

				visualizer.OnMapVisualizerStateChanged += (ModuleState state) => { 
					Debug.Log("Visualizer: " + state.ToString()); 
				};

				tileProvider.OnTileAdded += (UnwrappedTileId tileId) => { 
					Debug.Log("New tile"); 
				};
			};
		}

	}	
}
```

Clear the cache in the Mapbox menu and play the scene. You should get something similar to the image below, which is perfectly fine and as expected. We see that the visualizer starts working, twelve tiles are added and the visualizer finishes.

![screenshot_1](https://user-images.githubusercontent.com/5920351/42203751-1031c360-7ea0-11e8-857d-b30872acdf4d.png)

Now stop and play the scene again. The cache will now be used and we get something completely different. For each tile added the visualizer switches state six times, which means it finishes three times per tile. Which is equivalent to the three factories relating to raster, vector and height data.

![screenshot_2](https://user-images.githubusercontent.com/5920351/42203865-5bc9c49e-7ea0-11e8-8795-aaf05f48da46.png)

We can also see an inconsistency if we collapse the log. We see that twelve tiles are added but the visualizer only finishes eleven times. It appears the visualizer never changes state when the last tile is loaded.
![screenshot_3](https://user-images.githubusercontent.com/5920351/42203868-60d7caf8-7ea0-11e8-98de-eac9990f79ec.png)
![screenshot_4](https://user-images.githubusercontent.com/5920351/42203960-a615d7ea-7ea0-11e8-9fb1-4fc6c7c4aab3.png)

In summary, the cache messes up the intended behavior of the callbacks as it is not asynchronous and will not add the necessary delay for the visualizer to stay busy until all tiles are done. My suggestion is to involve the TileProvider and query its state or number of tiles that it can provide. 

I bypassed this issue myself by listening to TileAdded-events and simply keep track of all tiles, their raster states and if the tileprovider has supplied all tiles yet. (a simple count calculation was done by querying the options for the tileprovider)

**Comment by BergWerkGIS** [comment]
@DennyLindberg thanks for the detailed analysis 🙏 (I wish all tickets were like yours).

We are aware of this issue and already started overhauling tile fetching/caching/tile events/tile counts/...

Unfortunately it might take till end of summer until these changes will make it into a release as we are preparing the switch to NetStandard2.0 backend at the same time.

Nevertheless we'll discuss internally if we "borrow" 😏 some of your ideas for a short term solution.
/cc @atripathi-mb @isiyu

**Comment by DennyLindberg** [comment]
@BergWerkGIS Sounds good! There is no rush on my end as I have already circumvented the issue for our project. The SDK is already quite solid considering its early state, I expect the next major update to be quite an improvement.

**Comment by brnkhy** [comment]
@DennyLindberg admittedly I thought we fixed that spam issue (even though the map state events are still not that good).
At the moment I'm working on something that'll hopefully fix all these problems. What I'm trying is to change factories so they'll save tiles (requests) instead of executing them right away. And then an update function checks for saved tile requests and executes them whenever it sees fit.
This means at any given time, you'll be able to check how many tiles are in the queue, how many tiles are requested and waiting for response etc. 
Also this'll enable scheduling and workload distribution (over frames) and you'll be able to customize it such that it'll `request max of X tiles at a time` or `request only every T seconds` etc.
I just started this and first tests are going pretty good. As @BergWerkGIS said it might not make it into dev branch soon but I think there'll be some public test branchs soon.

cc @isiyu

**Comment by DennyLindberg** [comment]
@brnkhy That sounds like a much better approach. Looking forward to it.

**Comment by brnkhy** [comment]
these events should be fixed in v1.4.5, closing it for now. thanks again @DennyLindberg !
